Propagate Clang function types through SIL

varungandhi-apple opened the pull request:
Re-landing the patch from https://github.com/apple/swift/pull/29691.

There are still some issues that need to be ironed out:

- [x] ~~During type lowering, the Clang type is not being computed when we make an `@convention(c)` function.  As far as I know, this is the only place missing, but I need to investigate more.~~ [Updated the primary function in type lowering which does the adjustment, but there still seem to be spots which haven't been fixed..] We see an assertion failure in IRGen where the remangling check fails (the original type does not store a Clang type but the demangled type does).
- [x] I'm not sure if the changes from isDerivable bit PR https://github.com/apple/swift/pull/29349 should be also included in this PR or land that separately. There are a couple of possible strategies: (1) land this PR, land the isDerivable bit PR, land the groundwork for the C++ template importing or (2) land this PR, land the groundwork for the C++ template importing, land the isDerivable bit PR. (The isDerivable bit is necessary so that the name mangling for `@convention(c)` functions which have a derivable C type doesn't change).

More generally, I feel like the API around `ExtInfo` is a bit problematic; there is no good place to enforce invariants such as if `rep == CFunctionPointer || rep == Block`, then you must have an associated Clang type. The presence of the `with` methods means that you are sometimes forced to create illegal intermediate states when creating new `ExtInfo` values. I'm wondering if we should refactor it to use some kind of "builder" API where you have an `ExtInfoBuilder` which represents a WIP state, and then you call a `finish : (ExtInfoBuilder) -> ExtInfo` method which checks if the builder is well-formed, before emitting an `ExtInfo`.

varungandhi-apple commented on lib/AST/Decl.cpp:
```diff
@@ -2688,6 +2688,7 @@ mapSignatureExtInfo(AnyFunctionType::ExtInfo info,
     return AnyFunctionType::ExtInfo();
   return AnyFunctionType::ExtInfo()
       .withRepresentation(info.getRepresentation())
+      .withClangFunctionType(info.getClangFunctionType())
```
This is a bit dissatisfying, I'm considering having a method on ExtInfo to handle this, so we can avoiding having a `getClangFunctionType()` method.

varungandhi-apple commented:
Updated the primary function in type lowering which does the adjustment, but there still seem to be spots which haven't been fixed, still seeing the remangling check crash.

varungandhi-apple commented:
> The presence of the with methods means that you are sometimes forced to create illegal intermediate states when creating new ExtInfo values. 

This also manifests as us having code like the following, because we are not enforcing the invariant `rep == CFunctionPointer ==> NoEscape = true`.

```
static bool isEscaping(Type type) {
  if (auto *funcType = type->getAs<AnyFunctionType>()) {
    if (funcType->getExtInfo().getRepresentation() ==
          FunctionTypeRepresentation::CFunctionPointer)
      return false;

    return !funcType->getExtInfo().isNoEscape();
  }

  return false;
}
```

This is... not great IMO. If I see the `isNoEscape()` method, it seems reasonable to expect it to return the right value, regardless of what the representation is.

varungandhi-apple commented:
We should land the refactoring in https://github.com/apple/swift/pull/33085 first, assuming no one has major objections.

varungandhi-apple commented:
To give an update here:

1. I've broken out the first two commits into a separate PR #33541, along with some cleanup.
2. Asserting that a function type with `@convention(c)` or `@convention(block)` always carries a `clang::Type` has been trickier than expected; I will update this PR once that is done. At the moment, I'm changing parts of lowering to use `SILExtInfoBuilder` instead of an `ASTExtInfo` with a SIL representation (because the `isPseudogeneric` bit is computed right before creating the Clang type, but the rest of the ExtInfo including a SIL representation needs to be passed in to `getSILFunctionType`). However there are still issues to be ironed out due to the convention not being set correctly in certain places during lowering.

varungandhi-apple commented:
I ended up rebasing the base PR #33541, force-pushing and then started rebasing this PR. The conflicts that came up are coming up primarily due to changes to `SILFunctionType` (it is storing an extra async bit) in #33554 and related PRs. It would be better to fix that bit first and move it to `SILExtInfoBuilder` (I've pointed this out in a comment in #33554) and then rebase this PR.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
I wonder why this fails only on Linux. Not seeing a similar issue building SwiftPM on macOS, or with small snippets using higher-order functions on macOS.

<details>

<pre>
Incorrect reconstructed type for $sSvSgA3ASbIetCyyd_SgSbIetCyyyd_SgD
Original type:
(bound_generic_enum_type decl=Swift.(file).Optional
  (sil_function_type type=@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>, Optional<@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>) -> Bool>) -> Bool
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (sil_function_type type=@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>) -> Bool
        (input=bound_generic_enum_type decl=Swift.(file).Optional
          (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
        (input=bound_generic_enum_type decl=Swift.(file).Optional
          (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
        (result=struct_type decl=Swift.(file).Bool)
        (substitution_map generic_signature=<nullptr>)
        (substitution_map generic_signature=<nullptr>) clang_type=PointerType 0xb52e540 '_Bool (*)(void *, void *)'
`-FunctionProtoType 0xb52e490 '_Bool (void *, void *)' cdecl
  |-BuiltinType 0x924bb70 '_Bool'
  |-PointerType 0x924c310 'void *' imported
  | `-BuiltinType 0x924bb50 'void'
  `-PointerType 0x924c310 'void *' imported
    `-BuiltinType 0x924bb50 'void'
))
    (result=struct_type decl=Swift.(file).Bool)
    (substitution_map generic_signature=<nullptr>)
    (substitution_map generic_signature=<nullptr>)))
Reconstructed type:
(bound_generic_enum_type decl=Swift.(file).Optional
  (sil_function_type type=@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>, Optional<@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>) -> Bool>) -> Bool
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
    (input=bound_generic_enum_type decl=Swift.(file).Optional
      (sil_function_type type=@convention(c) (Optional<UnsafeMutableRawPointer>, Optional<UnsafeMutableRawPointer>) -> Bool
        (input=bound_generic_enum_type decl=Swift.(file).Optional
          (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
        (input=bound_generic_enum_type decl=Swift.(file).Optional
          (struct_type decl=Swift.(file).UnsafeMutableRawPointer))
        (result=struct_type decl=Swift.(file).Bool)
        (substitution_map generic_signature=<nullptr>)
        (substitution_map generic_signature=<nullptr>) clang_type=PointerType 0xb52e540 '_Bool (*)(void *, void *)'
`-FunctionProtoType 0xb52e490 '_Bool (void *, void *)' cdecl
  |-BuiltinType 0x924bb70 '_Bool'
  |-PointerType 0x924c310 'void *' imported
  | `-BuiltinType 0x924bb50 'void'
  `-PointerType 0x924c310 'void *' imported
    `-BuiltinType 0x924bb50 'void'
))
    (result=struct_type decl=Swift.(file).Bool)
    (substitution_map generic_signature=<nullptr>)
    (substitution_map generic_signature=<nullptr>) clang_type=PointerType 0xb4e82c0 '_Bool (*)(void *, void *, _Bool (*)(void *, void *))'
`-FunctionProtoType 0xb4e8210 '_Bool (void *, void *, _Bool (*)(void *, void *))' cdecl
  |-BuiltinType 0x924bb70 '_Bool'
  |-PointerType 0x924c310 'void *' imported
  | `-BuiltinType 0x924bb50 'void'
  |-PointerType 0x924c310 'void *' imported
  | `-BuiltinType 0x924bb50 'void'
  `-PointerType 0xb52e540 '_Bool (*)(void *, void *)'
    `-FunctionProtoType 0xb52e490 '_Bool (void *, void *)' cdecl
      |-BuiltinType 0x924bb70 '_Bool'
      |-PointerType 0x924c310 'void *' imported
      | `-BuiltinType 0x924bb50 'void'
      `-PointerType 0x924c310 'void *' imported
        `-BuiltinType 0x924bb50 'void'
))
</pre>
</details>

varungandhi-apple commented:
Did manage to reproduce the issue on Linux. The problem is coming up due to https://github.com/apple/swift-package-manager/blob/96b0ff907fdd4d91fcd207f4c49aae4fce1d9ec5/swift-tools-support-core/Sources/TSCUtility/IndexStore.swift#L215-L215 . We end up specializing a dlsym wrapper (!) with the generic parameter having type `(@convention(c) (UnsafeMutableRawPointer?, UnsafeMutableRawPointer?, (@convention(c) (UnsafeMutableRawPointer?, UnsafeMutableRawPointer?) -> Bool)?) -> Bool)?` thanks to
https://github.com/apple/indexstore-db/blob/8c593c430c714aac7e777b501f26e8e36849315e/include/indexstore/indexstore_functions.h#L539-L542

Not entirely sure _what_ is dropping the Clang type though, haven't been able to create a minimal example with the same type on macOS. 🤔

varungandhi-apple commented:
OK, there are at least two issues here.

1. We are not calling `checkInvariants()` in all the right constructors.
2. `UseClangFunctionTypes` really needs to be turned on at this point. Either that, or the mangling code needs to be relaxed or we add a temporary re-computation step that needs to be removed later. Otherwise, if the setting is false, we end up dropping Clang function types when computing canonical AST function types, which means that when we create a SIL ExtInfo for a pointer to a higher-order function, we may no longer have the Clang type for a parameter.

Fixing those two things leads to crashes elsewhere from the little that I tested, so there's probably a third problem somewhere.

Still not entirely clear why we hit this on Linux but not on macOS. 🤷🏼

varungandhi-apple committed:
[NFC] Add missing invariant checks for ExtInfoBuilders.

varungandhi-apple committed:
[Printer] Conditionally print Clang types in emitted SIL.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test Windows

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please clean test macOS

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please test Windows

varungandhi-apple commented:
@swift-ci please clean test macOS

varungandhi-apple commented:
`stdlib/Random.swift` is failing with on 32-bit iOS simulator.

<details>
<pre>
******************** TEST 'Swift(iphonesimulator-i386) :: stdlib/Random.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-iphonesimulator-i386/stdlib/Output/Random.swift.tmp" && mkdir -p "/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-iphonesimulator-i386/stdlib/Output/Random.swift.tmp" && xcrun --toolchain default --sdk '/Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk' /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/bin/swiftc -target i386-apple-ios7.0-simulator  -module-cache-path '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/i386-apple-ios7.0-simulator/clang-module-cache' -F '/Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks' -toolchain-stdlib-rpath  -Xlinker -rpath -Xlinker /usr/lib/swift  -swift-version 4  -Xfrontend -ignore-module-source-info  -module-cache-path '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/i386-apple-ios7.0-simulator/clang-module-cache' /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/validation-test/stdlib/Random.swift -o /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-iphonesimulator-i386/stdlib/Output/Random.swift.tmp/a.out -module-name main -Xfrontend -disable-access-control  && codesign -f -s - /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-iphonesimulator-i386/stdlib/Output/Random.swift.tmp/a.out && /usr/bin/env DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/iphonesimulator' LD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/iphonesimulator:' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/iphonesimulator' xcrun --toolchain default --sdk '/Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk' simctl spawn --standalone 'iPhone 5'  /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-iphonesimulator-i386/stdlib/Output/Random.swift.tmp/a.out
--
Exit Code: 254

Command Output (stderr):
--
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/validation-test/stdlib/Random.swift:28:9: warning: variable 'g' was never mutated; consider changing to 'let' constant
    var g = SystemRandomNumberGenerator()
    ~~~ ^
    let
<unknown>:0: error: fatal error encountered while reading from module 'Swift'; please file a bug report with your project and the crash log
<unknown>:0: note: module 'Swift' full misc version is '5.3(5.3)/Apple Swift version 5.3-dev (LLVM 19447e27024a8d4, Swift 19032d59c06c27b)'
<unknown>:0: note: compiling as Swift 4.1.50, with 'Swift' built as Swift 5.3 (this is supported but may expose additional compiler issues)

*** DESERIALIZATION FAILURE (please include this section in any bug report) ***
(see "While..." info below)
Stack dump:
0.	Program arguments: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/bin/swift-frontend -frontend -c -primary-file /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/validation-test/stdlib/Random.swift -target i386-apple-ios7.0-simulator -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -module-cache-path /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swift-macosx-x86_64/swift-test-results/i386-apple-ios7.0-simulator/clang-module-cache -swift-version 4 -ignore-module-source-info -disable-access-control -target-sdk-version 14.0 -module-name main -o /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/Random-081e9b.o 
1.	Apple Swift version 5.3-dev (LLVM 19447e27024a8d4, Swift 19032d59c06c27b)
2.	While evaluating request ExecuteSILPipelineRequest(Run pipelines { Mandatory Diagnostic Passes + Enabling Optimization Passes } on SIL for main.main)
3.	While running pass #425 SILModuleTransform "MandatorySILLinker".
4.	While deserializing SIL function "swift_stdlib_random"
0  swift-frontend           0x000000010bce7b75 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift-frontend           0x000000010bce6b48 llvm::sys::RunSignalHandlers() + 248
2  swift-frontend           0x000000010bce8156 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007fff6f1da5fd _sigtramp + 29
4  libsystem_platform.dylib 0x00007ffee844a9b0 _sigtramp + 18446744071447184336
5  libsystem_c.dylib        0x00007fff6f0b0808 abort + 120
6  swift-frontend           0x000000010817bdfe swift::ModuleFileSharedCore::fatal(llvm::Error) + 110
7  swift-frontend           0x0000000108121eef swift::ModuleFile::fatal(llvm::Error) + 63
8  swift-frontend           0x0000000108174eac swift::ModuleFile::fatal() + 60
9  swift-frontend           0x000000010815332b swift::SILDeserializer::readSILFunctionChecked(llvm::PointerEmbeddedInt<unsigned int, 31>, swift::SILFunction*, llvm::StringRef, bool, bool) + 8283
10 swift-frontend           0x0000000108166500 swift::SILDeserializer::lookupSILFunction(swift::SILFunction*, bool) + 368
11 swift-frontend           0x00000001081e85b5 swift::SerializedSILLoader::lookupSILFunction(swift::SILFunction*, bool) + 117
12 swift-frontend           0x0000000108213e08 swift::SILLinkerVisitor::process() + 472
13 swift-frontend           0x0000000108213c02 swift::SILLinkerVisitor::processFunction(swift::SILFunction*) + 82
14 swift-frontend           0x000000010827a702 swift::SILModule::linkFunction(swift::SILFunction*, swift::SILModule::LinkingMode) + 114
15 swift-frontend           0x0000000107fe2dae (anonymous namespace)::SILLinker::run() + 78
16 swift-frontend           0x0000000107ea649e swift::SILPassManager::runModulePass(unsigned int) + 558
17 swift-frontend           0x0000000107eaaeba swift::SILPassManager::execute() + 666
18 swift-frontend           0x0000000107ea2af8 swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 72
19 swift-frontend           0x0000000107ea2a93 swift::ExecuteSILPipelineRequest::evaluate(swift::Evaluator&, swift::SILPipelineExecutionDescriptor) const + 51
20 swift-frontend           0x0000000107ec5dcd swift::SimpleRequest<swift::ExecuteSILPipelineRequest, std::__1::tuple<> (swift::SILPipelineExecutionDescriptor), (swift::RequestFlags)1>::evaluateRequest(swift::ExecuteSILPipelineRequest const&, swift::Evaluator&) + 29
21 swift-frontend           0x0000000107ead57b llvm::Expected<swift::ExecuteSILPipelineRequest::OutputType> swift::Evaluator::getResultUncached<swift::ExecuteSILPipelineRequest>(swift::ExecuteSILPipelineRequest const&) + 363
22 swift-frontend           0x0000000107ea2d24 swift::executePassPipelinePlan(swift::SILModule*, swift::SILPassPipelinePlan const&, bool, swift::irgen::IRGenModule*) + 68
23 swift-frontend           0x0000000107eb04b7 swift::runSILDiagnosticPasses(swift::SILModule&) + 87
24 swift-frontend           0x000000010790f488 swift::CompilerInstance::performSILProcessing(swift::SILModule*) + 56
25 swift-frontend           0x0000000107822b1e performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*) + 894
26 swift-frontend           0x000000010782245e performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 350
27 swift-frontend           0x000000010781a8ed swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4589
28 swift-frontend           0x00000001077b510d main + 861
29 libdyld.dylib            0x00007fff6efe1cc9 start + 1
<unknown>:0: error: unable to execute command: Abort trap: 6
<unknown>:0: error: compile command failed due to signal 6 (use -v to see invocation)
</pre>
</details>

varungandhi-apple commented:
For 32-bit iOS/watchOS, both `int` and `long` are 32-bits. So the failure happens in

https://github.com/apple/swift/blob/bfc9b5447f4499bfb363b948b91524ad6e5759a7/lib/Serialization/DeserializeSIL.cpp#L589-L594

The deserialized type `ty` ends up having storing a `long` corresponding a Swift `Int` argument, whereas the existing function `fn` ends up having an `int`. The Clang types end up being different, so the Swift types end up being different. 😬

varungandhi-apple commented:
Reduced test case:

```
// RUN: %target-swift-frontend %s -emit-ir -o /dev/null

import StdlibUnittest
import SwiftShims
let _ = swift_stdlib_random
```

varungandhi-apple commented:
Looking at a slightly less complicated example.

```
// RUN: %target-swift-frontend %s -emit-ir -o /dev/null

import StdlibUnittest
import SwiftShims

// void swift_stdlib_random(void *buf, __swift_size_t nbytes);
func f(_ p: UnsafeMutableRawPointer, _ c: Int) { swift_stdlib_random(p, c) }
```

In `applyNormalCall`:

```
RValue CallEmission::applyNormalCall(SGFContext C) {
  // We use the context emit-into initialization only for the
  // outermost call.
  SGFContext uncurriedContext = C;

  auto formalType = callee.getSubstFormalType();
  // ^ @convention(thin) (UnsafeMutableRawPointer, Int) -> ()
  auto origFormalType = callee.getOrigFormalType();
  // ^ @convention(c, cType: "void (*)(void * _Nonnull, __swift_size_t)") (UnsafeMutableRawPointer, Int) -> ()

  bool isCurried = (uncurriedSites.size() < callee.getParameterListCount());
  assert(!isCurried);

  // Get the callee type information.
  auto calleeTypeInfo = callee.getTypeInfo(SGF);
  // ^ calleeTypeInfo.substFnType = @convention(c, cType: "void (*)(void *, int)") (UnsafeMutableRawPointer, Int) -> ()
  // BAD! We shouldn't have dropped the info from the original formal type and only
  // to recompute the information from the Swift type later on.
```

Looking at `getTypeInfo`, it seems like something is off with the `SILDeclRef Constant` being stored in the `Callee` class instance.

We end up creating a `SILFunction fn` for `swift_stdlib_random` with type `@convention(c, cType: "void (*)(void *, int)") (UnsafeMutableRawPointer, Int) -> ()` thanks to the above code. This gets passed into SIL deserialization to update the function's linkage. However, during deserialization, we end up getting a `void (*)(void *, long)` Clang type for this function. This is strange because:
   - `__swift_size_t` is specially imported into Swift as `Int` according to `MappedTypes.def`.
  -  `__swift_size_t` is defined as a typedef to (essentially) `long unsigned int`. Notice the `unsigned`. But somehow the `unsigned` is gone? 🤔

Due to this mismatch, deserialization falls over.

rjmccall commented:
Why is deserialization falling over?  The import mapping is definitely not 1-1; certain unsigned types being imported as Int is the least of the problems there.  But I don't know why the type would be inconsistent in different sources.

varungandhi-apple commented:
Deserialization falls over because, at the moment, the types are expected to match up exactly. It's not clear to me what ought to be done in case of a mismatch here.

https://github.com/apple/swift/blob/bfc9b5447f4499bfb363b948b91524ad6e5759a7/lib/Serialization/DeserializeSIL.cpp#L589-L594

> The import mapping is definitely not 1-1; certain unsigned types being imported as Int is the least of the problems there.

Yeah, I'm not saying that's a problem. But we need to maintain a consistent view internally. I think something is off with this part I highlighted earlier:

```
  auto formalType = callee.getSubstFormalType();
  // ^ @convention(thin) (UnsafeMutableRawPointer, Int) -> ()
  auto origFormalType = callee.getOrigFormalType();
  // ^ @convention(c, cType: "void (*)(void * _Nonnull, __swift_size_t)") (UnsafeMutableRawPointer, Int) -> ()

  auto calleeTypeInfo = callee.getTypeInfo(SGF);
  // ^ calleeTypeInfo.substFnType = @convention(c, cType: "void (*)(void *, int)") (UnsafeMutableRawPointer, Int) -> ()
```

It's not clear to me why the substituted formal type is `@convention(thin)` instead of `@convention(c)`. If that's necessary, then we need a way to make sure that `calleeTypeInfo.substFnType` ends up using the Clang type from the original formal type.

rjmccall commented:
SIL type lowering definitely needs to be using the Clang type from the original formal type if there is one.

varungandhi-apple commented:
So I misspoke slightly here. The original formal type somehow ends up being `@convention(thin, cType: "void (*)(void * _Nonnull, __swift_size_t)")`, which should be impossible. The code for creating an ASTExtInfo value is very straight-forward... and I don't see anything off there. Going to try to use ASan+UBSan to see if there's any memory corruption...

varungandhi-apple commented:
OK. I realized my mistake. I was looking at an _AbstractionPattern_, not an `@convention(thin)` function carrying a Clang type. It is possible for an abstraction pattern to have an `@convention(thin)` function and also simultaneously carry a Clang type.

The problem (or at least one of them) was that when we create the SILConstantInfo, the formal and lowered types end up being `@convention(thin)`, but the SILFnType ends up being `@convention(c)`. However, we were passing in the ExtInfo through the lowered type for creating the SILFnType, instead of using the one in `bridgedTypes.Pattern`.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please clean test

varungandhi-apple commented:
The mangling check needs to be temporarily loosened until we start including the Clang type in the mangling.

varungandhi-apple commented:
I guess I could temporarily add an AST visitor that only overrides the *FunctionType cases to replace the Clang type with a dummy placeholder and use that to compare things.

The "sledgehammer" approach to this would just set `DisableRoundTripDebugTypes = true` until I've landed the mangling change, but I'm wary of leaving it on like that for potentially several weeks. 😬

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
The macOS failure is due to trying to compile a benchmark involving Float80 on armv7. https://github.com/apple/swift/pull/33801

varungandhi-apple commented on tools/sil-func-extractor/SILFunctionExtractor.cpp:
```diff
@@ -340,7 +343,7 @@ int main(int argc, char **argv) {
     auto SILOpts = SILOptions();
     SILOpts.EmitVerboseSIL = EmitVerboseSIL;
-    SILOpts.EmitSortedSIL = EnableSILSortOutput;
+    SILOpts.EmitSortedSIL = EmitVerboseSIL;
```
This seems off. Why would I change this?

varungandhi-apple commented:
@swift-ci please test macOS

varungandhi-apple committed:
[SIL] Store ClangTypeInfo in SILFunctionType.

This patch includes a large number of changes to make sure that:
1. When ExtInfo values are created, we store a ClangTypeInfo if applicable.
2. We reduce dependence on storing SIL representations in ASTExtInfo values.
3. Reduce places where we sloppily create ASTExtInfo values which should
   store a Clang type but don't. In certain places, this is unavoidable;
   see [NOTE: ExtInfo-Clang-type-invariant].

Ideally, we would check that the appropriate SILExtInfo does always store
a ClangTypeInfo. However, the presence of the HasClangFunctionTypes option
means that we would need to condition that assertion based on a dynamic check.
Plumbing the setting down to SILExtInfoBuilder's checkInvariants would be too
much work. So we weaken the check for now; we should strengthen it once we
"turn on" HasClangFunctionTypes and remove the dynamic feature switch.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
I think this is ready for review @rjmccall.

1. I've left the `EqualUpToClangTypes` check in as a stop-gap measure to prevent the compiler from crashing while still having round-trip testing when building with `UseClangFunctionTypes` turned on. I intend on removing it in a follow-up PR which adds the mangling.
2. I have some basic tests in place, should I add more test cases or do you think this is sufficient?
3. There are a bunch of formatting changes due to `clang-format`, which make this change larger than it ought to be. 🤷🏼

varungandhi-apple commented:
@swift-ci please test source compatibility

rjmccall commented on include/swift/AST/ExtInfo.h:
```diff
@@ -81,6 +83,26 @@ class ClangTypeInfo {
+    bool expectNonnullForCOrBlock, bool expectCanonical);
+
+  void dump();
+};
```
This seems like a lot of abstraction for an assertion, but if you think it's necessary, we can roll with it.

varungandhi-apple replied:
Yeah, I'm not a big fan of it, but it seemed like the cleanest approach given that we need to check for the error in three different places, including this one in `SILFunctionType::get(`.

```
  if (ctx.LangOpts.UseClangFunctionTypes) {
    if (auto error = ext.checkClangType()) {
      error.getValue().dump();
      llvm_unreachable("Unexpected Clang type in SILExtInfo.");
    }
  }
```

rjmccall commented on include/swift/AST/PrintOptions.h:
```diff
@@ -594,7 +594,7 @@ struct PrintOptions {
 
   /// Retrieve the set of options suitable for printing SIL functions.
-  static PrintOptions printSIL() {
+  static PrintOptions printSIL(bool printFullConvention = false) {
```
Is there something more general we should be passing in here, like a `SILOptions`?

varungandhi-apple replied:
We could pass an `Optional<SILPrintContext>` argument and project one of its fields (right now, some callers are doing that), it's not the perfect fit IMO (it carries a bunch of stuff which isn't useful) but we can use it.

varungandhi-apple replied:
But `SILPrintContext` lives in SIL, so maybe we don't want to depend on it in the AST.

rjmccall replied:
We can declare the method in the AST headers and then implement it in the SIL library somewhere; this wouldn't be the first time we did that.

varungandhi-apple replied:
Ok. I already punted some NFC changes to a follow-up PR, I'll address this bit there.

rjmccall commented on lib/AST/ExtInfo.cpp:
```diff
@@ -64,14 +54,99 @@ void ClangTypeInfo::dump(llvm::raw_ostream &os,
+    type->dump();
+    break;
+  }
+  llvm_unreachable("Unhandled case for UnexpectedClangTypeError");
```
This isn't the right place for an unreachable; it needs to go after the switch, to "cover" the implicit exit when a matching case isn't present in the switch.  Here it isn't doing anything.  This means that the cases need to `return` instead of `break`.

rjmccall commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -3065,26 +3057,47 @@ static bool isImporterGeneratedAccessor(const clang::Decl *clangDecl,
+    }
+    if (clangType) {
+      // According to [NOTE: ClangTypeInfo-contents], we need to wrap a function
+      // type in an additional clang::PointerType.
```
What is this NOTE markup?  Is this an editor feature we're assuming support for, or is it just a reference to the documentation?

varungandhi-apple replied:
It's a cross-reference to documentation elsewhere. It's not supported by any editors, it just helps to indicate coupling between two things. Generally, I prefer to document invariants etc. near the type definition/invariant check. So when you change the type definition/invariant, it is easier to track down what else is relying on it. Another benefit is that when you need to write a longer comment, you can write it out-of-line, which avoids obscuring the surrounding code. https://github.com/apple/swift/blob/master/lib/FrontendTool/FrontendTool.cpp has a bunch of similar named NOTEs because of the different edge cases in the module trace.

rjmccall reviewed (commented):
Sorry for the slow review.  Could you make a separate PR that just does the NFC set-up work, like converting various places to use `ExtInfoBuilder` or `ExtInfo::getThin()`?

varungandhi-apple commented on include/swift/AST/ExtInfo.h:
```diff
@@ -604,31 +694,41 @@ class SILExtInfoBuilder {
 
   // Note that we don't have setters. That is by design, use
   // the following with methods instead of mutating these objects.
+  LLVM_NODISCARD
```
In the past, the `withXYZ` methods on `AnyFunctionType::ExtInfo`and `SILFunctionType::ExtInfo` were marked `LLVM_NODISCARD` presumably because they were "immutable update methods", so it doesn't make sense to call the methods and ignore the result. There was an oversight by me when refactoring the code to use the builder-style API, where some `withXYZ` methods were marked `LLVM_NODISCARD` but not others. I was fixing that oversight.

rjmccall replied:
Ah, okay.

varungandhi-apple commented:
Fixed the bug in `UnexpectedClangTypeError::dump` and removed a couple of cleanup commits which can be added separately (will land those in a follow-up PR).

@swift-ci test

varungandhi-apple commented:
@swift-ci test

rjmccall reviewed (approved):
Generally looks good to me.

varungandhi-apple commented:
@swift-ci please test and merge

varungandhi-apple commented:
@swift-ci please test Windows

varungandhi-apple commented:
@swift-ci test macOS platform

varungandhi-apple commented:
@swift-ci test Linux platform

varungandhi-apple commented:
@swift-ci please clean test Windows
